### PR TITLE
Adding PaulBouchiers covariance changes

### PR DESCRIPTION
**Pull request by Nick-Sullivan**
Regarding covariance generation in Issue #17 
I've plucked the covariance changes made by Paul which never got merged (https://github.com/ros-drivers/nmea_navsat_driver/pull/19)

It uses GST covariance where available, otherwise uses default covariances estimated from fix type.

I've tested it on a ublox C94-M8P.

### CODE REVIEW

File: scripts/nmea_topic_serial_reader [hunk 1]
```diff
@@ -42,7 +42,7 @@ from libnmea_navsat_driver.driver import RosNMEADriver
 if __name__ == '__main__':
     rospy.init_node('nmea_topic_serial_reader')
 
-    nmea_pub = rospy.Publisher("nmea_sentence", Sentence)
+    nmea_pub = rospy.Publisher("nmea_sentence", Sentence, queue_size=1)
```
[the remark this answers is not in the document]
Comment (Nick-Sullivan) [reply scripts/nmea_topic_serial_reader]: I set it to 1 for consistency with the publishers in driver.py, but I agree they should all be something higher. Should I change them all or leave it for a separate PR?

File: src/libnmea_navsat_driver/driver.py [hunk 2]
```diff
@@ -110,17 +141,22 @@ def add_sentence(self, nmea_string, frame_id, timestamp=None):
...
             current_fix.longitude = longitude
 
-            hdop = data['hdop']
-            current_fix.position_covariance[0] = hdop ** 2
-            current_fix.position_covariance[4] = hdop ** 2
-            current_fix.position_covariance[8] = (2 * hdop) ** 2  # FIXME
-            current_fix.position_covariance_type = \
-                NavSatFix.COVARIANCE_TYPE_APPROXIMATED
-
-            # Altitude is above ellipsoid, so adjust for mean-sea-level
```
[the remark this answers is not in the document]
Comment (Nick-Sullivan) [reply src/libnmea_navsat_driver/driver.py]: Accident, I'll put it back in.

File: src/libnmea_navsat_driver/driver.py [hunk 3]
```diff
@@ -79,22 +91,41 @@ def add_sentence(self, nmea_string, frame_id, timestamp=None):
             current_time_ref.source = frame_id
 
         if not self.use_RMC and 'GGA' in parsed_sentence:
+            current_fix.position_covariance_type = \
+                NavSatFix.COVARIANCE_TYPE_APPROXIMATED
+              
             data = parsed_sentence['GGA']
             gps_qual = data['fix_type']
-            if gps_qual == 0:
+            if gps_qual == 0:      # No fix, reset covariance
```
[the remark this answers is not in the document]
Comment (Nick-Sullivan) [reply src/libnmea_navsat_driver/driver.py]: Sounds good to me

File: src/libnmea_navsat_driver/parser.py [hunk 4]
```diff
@@ -131,6 +141,8 @@ def parse_nmea_sentence(nmea_sentence):
                      % repr(nmea_sentence))
         return False
     fields = [field.strip(',') for field in nmea_sentence.split(',')]
+    last_two_fields = fields[-1].split('*')
```
[the remark this answers is not in the document]
Comment (Nick-Sullivan) [reply src/libnmea_navsat_driver/parser.py]: This was copied over from Paul's code. It removes the checksum, but the parser seems to work fine without it so I'll remove it

File: src/libnmea_navsat_driver/driver.py [hunk 5]
```diff
@@ -49,7 +49,19 @@ def __init__(self):
 
         self.time_ref_source = rospy.get_param('~time_ref_source', None)
         self.use_RMC = rospy.get_param('~useRMC', False)
-
+        
+        self.fix_type = 'invalid'
```
[the remark this answers is not in the document]
Comment (Nick-Sullivan) [reply src/libnmea_navsat_driver/driver.py]: From what I can see, the NavSatFix doesn't provide the enum's we need. Is there one in ROS that has {Invalid, SPS, DGPS, RTK Fix, RTK Float, WAAS, and Unknown}?

File: src/libnmea_navsat_driver/driver.py [hunk 6]
```diff
@@ -110,17 +141,22 @@ def add_sentence(self, nmea_string, frame_id, timestamp=None):
...
+                self.lon_std_dev = self.default_epe
+            if not self.using_receiver_epe or math.isnan(self.lat_std_dev):
+                self.lat_std_dev = self.default_epe
+            if not self.using_receiver_epe or math.isnan(self.alt_std_dev):
+                self.alt_std_dev = self.default_epe * 2
 
+            hdop = data['hdop']
+            current_fix.position_covariance[0] = (hdop * self.lon_std_dev) ** 2
+            current_fix.position_covariance[4] = (hdop * self.lat_std_dev) ** 2
+            current_fix.position_covariance[8] = (hdop * self.alt_std_dev) ** 2  # FIXME
```
[the remark this answers is not in the document]
Comment (Nick-Sullivan) [reply src/libnmea_navsat_driver/driver.py]: I'm not very knowledgeable on GPS covariance estimation, so I blindly copied this from Paul's code. I'm happy to change it back to scaling by 2.

File: src/libnmea_navsat_driver/driver.py [hunk 7]
```diff
@@ -79,27 +137,19 @@ def add_sentence(self, nmea_string, frame_id, timestamp=None):
...
+            fix_type = data['fix_type']
+            if not (fix_type in self.gps_qualities):
+              fix_type = -1
+            gps_qual = self.gps_qualities[fix_type]
+            default_epe = gps_qual[0];
+            current_fix.status.status = gps_qual[1]
+            current_fix.position_covariance_type = gps_qual[2];
             current_fix.status.service = NavSatStatus.SERVICE_GPS
 
-            current_fix.header.stamp = current_time
```
[the remark this answers is not in the document]
Comment (Nick-Sullivan) [reply src/libnmea_navsat_driver/driver.py]: It's a duplicate of line 71 (line 129 of this PR)

**Comment by Nick-Sullivan** [comment]
Done!

**Comment by evenator** [comment]
This change looks good to me, and I didn't observe any regressions testing with my USB device.

If no one objects before 21:00 UTC (5 PM US Eastern time) on Monday August 27, I'll merge this change.

I'm not sure if the change can be backported to current versions, because users may be depending on the existing covariance calculation. If the change isn't backported, it would be released in ROS N in Spring 2019. In the mean time, anyone who wants the improvement can build the package from source.

**Pull request merged**